### DISCUSSION TITLE
Endpoint indexing

### DISCUSSION THREAD

**Discussion by adirobu84** [sekigon-gonnoc/Pico-PIO-USB#95] [General]
I noticed that only the IN endpoints are indexed and the function `pio_usb_get_endpoint` is therefore only able to return IN endpoints. What about OUT endpoints? Is there a reason for this or simply an oversight?

In my code I expect to call the functions `pio_usb_get_in_data` and `pio_usb_set_out_data` to receive and send data from/to the host. For that I need a pointer to the respective endpoints, which are retrieved using `pio_usb_get_endpoint`. Not being able to get a pointer to an OUT endpoint, I'm not sure how `pio_usb_get_in_data` can be used.
For clarification, and OUT endpoint refers to direction from host to device (OUT from host) and we would use `pio_usb_get_in_data` to retrieve it on the device. Similarly `pio_usb_set_out_data` sends data on an IN endpoint, which transfers it from the device to the host (IN to host).

Now, I can just modify the API so I can get both IN and OUT endpoints using `pio_usb_get_endpoint`, but I am wondering if I am missing something and how others are using the API to implement their devices?

In general, I'm missing a good example on interacting with both control and regular endpoints. If anyone has some guidance this would be much appreciated.

**Comment by adirobu84** [comment]
A few more findings: The function `pio_usb_device_get_endpoint_by_address()` is a better option as it not only works for both IN and OUT endpoints, but also uses the endpoints actual address instead of an ominous index. The only downside is that this is not part of the regular API and requires the inclusion of `pio_usb_ll.h` (I'd say that's not really a downside, but it would be great if it was part of the top-level API).

So as far as endpoint indexing goes I think I have what I need, but I am still curious if `pio_usb_get_endpoint()` not supporting OUT endpoints is intentional or a bug?

**Comment by mikeDonahoe** [comment]
@adirobu84 I think pio_usb_device_transfer is what you are looking for.

**Reply by adirobu84** [reply]
Does it? Only for IN transfers (device to host) though. For OUT transfers (host to device) the data itself is already in the endpoint buffer and needs to be copied out of the buffer into an application buffer. That's not what pio_usb_device_transfer does. It only initiates an endpoint transfer.

My question was more on why the top-level application API is not consistent and is missing some functionality. Specifically, the function pio_usb_get_endpoint in said application API only works for IN endpoints.

The usb_device example, which emulates a moving mouse, uses pio_usb_get_endpoint and then pio_usb_set_out_data. Both of these functions are in pio_usb.c in the section labeled "Application API". This example works fine since set_out_data is setting up the IN endpoint for a transaction, but if someone were to extrapolate from this example and use pio_usb_get_in_data it wouldn't work since this operates on an OUT endpoint and pio_usb_get_endpoint only returns IN endpoints.

I think the "issue" here is simply an incomplete or scattered API and not having a good example. As mentioned in a previous comment, the function pio_usb_device_get_endpoint_by_address() works for both IN and OUT endpoints (but it's not part of the described application API. I've had success when using this function along with both pio_usb_get_in_data and pio_usb_get_out_data

I am happy to push a branch that fixes this and/or adds a more complete example, but before I do so I was simply wondering if I am missing something altogether.

**Reply by mikeDonahoe** [reply]
Hmm, I'm using pio_usb_device_transfer on a device in a task to receive data from the host.  See my earlier question #68. Nonetheless, I see your point as far as getting both IN/OUT endpoints and admittedly I went down the same rabbit hole when I was first attempting to process OUT transfers.
